ncm-ssh/src/main/pan/components/ssh/schema.pan

```diff
@@ -10,18 +10,63 @@ include 'pan/types';
 
 type ssh_preferred_authentication = string with match(SELF, '^(gssapi-with-mic|hostbased|publickey|keyboard-interactive|password)$');
 
+type ssh_ciphers = string with is_valid_ssh_cipher(SELF);
+type ssh_hostkeyalgorithms = string with match(SELF, "^(ssh-(rsa|dss|ed25519)|ecdsa-sha2-nistp(256|384|521)|(ssh-rsa-cert-v01|ssh-dss-cert-v01|ecdsa-sha2-nistp256-cert-v01|ecdsa-sha2-nistp384-cert-v01|ecdsa-sha2-nistp521-cert-v01|ssh-rsa-cert-v00|ssh-dss-cert-v00|ssh-ed25519-cert-v01)@openssh.com)$");
+type ssh_kbdinteractivedevices = string with match (SELF, "^(bsdauth|pam|skey)$");
+type ssh_kexalgorithms = string with match (SELF, "^(diffie-hellman-group-exchange-sha256|ecdh-sha2-nistp(256|384|521)|curve25519-sha256@libssh.org)$");
+type ssh_MACs = string with is_valid_ssh_MAC(SELF);
+
+function is_valid_ssh_MAC = {
+    match(ARGV[0], "^(hmac-(sha1|sha2-256|sha2-512|ripemd160)|(hmac-ripemd160|umac-64|umac-128|hmac-sha1-etm|hmac-sha2-256-etm|hmac-sha2-512-etm|hmac-ripemd160-etm|umac-64-etm|umac-128-etm)@openssh.com)$");
+};
+
+function is_valid_ssh_cipher = {
+    match (ARGV[0], "^((aes128|aes192|aes256)-ctr|(aes128-gcm|aes256-gcm|chacha20-poly1305)@openssh.com)$");
+};
+
+function is_valid_ssh_kexalgorithm = {
+    match (ARGV[0], "^(diffie-hellman-group-exchange-sha256|ecdh-sha2-nistp(256|384|521)|curve25519-sha256@libssh.org)$");
+};
+
+type legacy_ssh_MACs = string with {
+    foreach(i; cmp; split(',', SELF)){
+        if(!is_valid_ssh_MAC(cmp)){
+            error("Invalid or insecure MAC found:" + cmp);
+        };
+    };
+    true;
+};
+
+type legacy_ssh_ciphers = string with {
+    foreach(i; cmp; split(',', SELF)){
+        if(!is_valid_ssh_cipher(cmp)){
+            error("Invalid or insecure cipher found:" + cmp);
+        };
+    };
+    true;
+};
+
+type legacy_ssh_kexalgorithm = string with {
+    foreach(i; cmp; split(',', SELF)){
+        if(!is_valid_ssh_cipher(cmp)){
+            error("Invalid or insecure key exchange algorithm found:" + cmp);
+        };
+    };
+    true;
+};
+
 type ssh_core_options_type = {
     "AddressFamily" ? string with match (SELF, '^(any|inet6?)$')
     "ChallengeResponseAuthentication" ? legacy_binary_affirmation_string
-    "Ciphers" ? string
+    "Ciphers" ? legacy_ssh_ciphers
```

Comment:
wel, it used to be string, so we can't just change it into a list? that's why I had to create the legacy string with...